Optimize redundant logic used in queries

pdiaz opened the pull request:
After upgrading parse-server from 2.7.4 to 4.4.0, MongoDB 4.0 maxed CPU usage from what used to be barely 20% usage.

This does reduce the redundant complexity on queries that involve pointer permissions.

dplewis commented:
Thanks for the PR? Can you write a test case for this?

mtrezza commented:
Would you please also open an issue for this to explain the proposed change? This allow others to join the discussion more easily and helps us to better trace changes in Parse Server.

mtrezza commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ### master
 [Full Changelog](https://github.com/parse-community/parse-server/compare/4.4.0...master)
+- IMPROVE: Optimize queries on classes with pointer permissions.
```
Add: `[#7061](https://github.com/parse-community/parse-server/pull/7061). Thanks to [Pedro Diaz](https://github.com/pdiaz)`

You can remove your name of course if you prefer to do so.

mtrezza commented on src/Controllers/DatabaseController.js:
```diff
@@ -1365,6 +1365,32 @@ class DatabaseController {
+      }
+    }
+
+    if (query.$or.length === 1) return query.$or[0];
```
Can you please wrap the if block in brackets?

@dplewis Shouldn't lint currently enforce this? If not, I think we should add that rule; if blocks without brackets are just an unnecessary bug risk IMHO.

mtrezza commented on src/Controllers/DatabaseController.js:
```diff
@@ -1365,6 +1365,32 @@ class DatabaseController {
+    }
+
+    if (query.$or.length === 1) return query.$or[0];
+
```
Please remove empty line.

mtrezza commented on src/Controllers/DatabaseController.js:
```diff
@@ -1448,13 +1474,23 @@ class DatabaseController {
+          if (foundEntries === queryClauseEntries.length) {
+            return query;
+          }
+
```
Please remove empty line.

mtrezza reviewed (changes requested):
Some minor comments

dplewis commented on src/Controllers/DatabaseController.js:
```diff
@@ -1365,6 +1365,65 @@ class DatabaseController {
+
+  reduceOrOperation(query: { $or: Array<any> }): any {
+    if (!query.$or) {
+      return query;
```
Can you add write test case for if `$or` doesn't exist?

dplewis commented on src/Controllers/DatabaseController.js:
```diff
@@ -1365,6 +1365,65 @@ class DatabaseController {
+
+  reduceAndOperation(query: { $and: Array<any> }): any {
+    if (!query.$and) {
+      return query;
```
Can you add write test case for if `$and` doesn't exist?

dplewis reviewed (changes requested):
Just a few nits

dplewis commented:
Have you tried your branch in production?

mtrezza commented:
Before merging, mind any ongoing discussion in https://github.com/parse-community/parse-server/issues/7065.

pdiaz commented:
> Have you tried your branch in production or tested it against your large database?

Yes, after this optimization performance went back to the same range as before.

dplewis reviewed (approved):
LGTM! @mtrezza How does it look?

mtrezza commented:
I see that the test cases test the new methods of the database controller, but there is no high-level test case that tests whether these methods are implemented correctly within the controller. Can we add a test case that makes a Parse.Query and evaluates the optimized query that would be sent to the DB client?

pdiaz commented:
> I see that the test cases test the new reduction methods of the database controller, but there is no high-level test case that tests whether these methods are implemented correctly within the controller and in which scenarios they are actually applied to a query. Can we add some test cases that make a `ParseQuery` and ensure the optimization of the resulting query?

Good call! I previously tested only with two pointer permissions, so I added two extra test cases for 2 and 3 pointer permissions on a class and fixed and issue detected with the second case.

mtrezza reviewed (changes requested):
Just trivia, but on a general note:

Does this optimization apply to all $or / $and across a query or only related pointer permissions?

pdiaz commented on spec/DatabaseController.spec.js:
```diff
@@ -236,6 +236,65 @@ describe('DatabaseController', function () {
+        query,
+        ACL_GROUP
+      );
+
```
All empty lines? If not between what exactly?

pdiaz commented:
> Does this optimization apply to all $or / $and across a query or only related pointer permissions?

@mtrezza No all, only on find operations related to pointer permissions and not the queries themselves. Update operations could also have unoptimized $and operations.

A full optimizer for queries in general would require many more test cases.

mtrezza commented:
> A full optimizer for queries in general would require many more test cases.

I was thinking into that direction. A Parse Query Optimizer would be an interesting feature, but maybe that's more a job for the guys at MongoDB.

This PR looks good to me, ready for merge I'd say.

pdiaz commented:
> > A full optimizer for queries in general would require many more test cases.
> 
> I was thinking into that direction. A Parse Query Optimizer would be an interesting feature, but maybe that's more a job for the guys at MongoDB.
> 
> This PR looks good to me, ready for merge I'd say.

Leaving that task to MongoDB guys means that a MongoDB upgrade would be required for everybody and that would take even much longer. I've been already working on that idea in my own, it will take some time to be ready. 

This PR right now is already a good step forward.

pdiaz commented:
@mtrezza @dplewis Looks like PG tests are a bit flaky. Can you trigger a restart?

mtrezza commented:
Checks passed. @dplewis any thoughts before merging?

parseplatformorg commented:
🎉 This change has been released in version [5.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.1)

parseplatformorg commented:
🎉 This change has been released in version [5.0.0](https://github.com/parse-community/parse-server/releases/tag/5.0.0)
